### PR TITLE
Develop v1.3

### PR DESCRIPTION
**Pull request by edwardivash**

### CODE REVIEW

File: RSReader/Categories/ErrorAlertController/UIAlertController+ErrorAlertController.m [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+//
+//  Created by Eduard Ivash on 3.01.21.
+//
+
+#import "UIAlertController+ErrorAlertController.h"
+#import "UIAlertController+RSSAlertController.h"
+
+@class RSSViewController;
+
+@implementation UIAlertController (ErrorAlertController)
```
Comment (OrientCue) [inline RSReader/Categories/ErrorAlertController/UIAlertController+ErrorAlertController.m:13]: There are a lot of repeated code here. Don't repeat yourself.
**DRY violation.**

File: RSReader/Categories/ErrorAlertController/UIAlertController+ErrorAlertController.m [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
++ (UIAlertController *)showErrorAlertUrlFormat {
+    UIAlertController *errorAlertController = [UIAlertController alertControllerWithTitle:@"Wrong URL format." message:@"Example: https://news.tut.by \nor https://tut.by/rss" preferredStyle:UIAlertControllerStyleAlert];
```
Comment (OrientCue) [inline RSReader/Categories/ErrorAlertController/UIAlertController+ErrorAlertController.m:16]: Please use localized strings for text which is displayed to user.

File: RSReader/Categories/ErrorAlertController/UIAlertController+ErrorAlertController.m [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"Ok" style:UIAlertActionStyleCancel handler:nil];
```
Comment (OrientCue) [inline RSReader/Categories/ErrorAlertController/UIAlertController+ErrorAlertController.m:17]: same

File: RSReader/Categories/ErrorAlertController/UIAlertController+ErrorAlertController.m [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    [errorAlertController addAction:okAction];
+    return errorAlertController;
+}
+
++(UIAlertController *)showErrorAlertDublicateUrl {
+    UIAlertController *errorAlertController = [UIAlertController alertControllerWithTitle:@"This channel already exists." message:@"Enter a new URL for a channel." preferredStyle:UIAlertControllerStyleAlert];
```
Comment (OrientCue) [inline RSReader/Categories/ErrorAlertController/UIAlertController+ErrorAlertController.m:23]: same

File: RSReader/Categories/ErrorAlertController/UIAlertController+ErrorAlertController.m [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    UIAlertAction *okAction = [UIAlertAction actionWithTitle:@"Ok" style:UIAlertActionStyleCancel handler:nil];
+    [errorAlertController addAction:okAction];
+    return errorAlertController;
+}
+
++(UIAlertController *)showAlertEmptyData:(UINavigationController *)nvc {
+    UIAlertController *errorAlertController = [UIAlertController alertControllerWithTitle:@"It is not possible to download data from this RSS feed channel." message:@"Please edit url." preferredStyle:UIAlertControllerStyleAlert];
```
Comment (OrientCue) [inline RSReader/Categories/ErrorAlertController/UIAlertController+ErrorAlertController.m:30]: same

File: RSReader/Categories/RSSAlertController/UIAlertController+RSSAlertController.m [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+#import "UIAlertController+ErrorAlertController.h"
+#import "FeedService.h"
+#import "HtmlParser.h"
+
+@implementation UIAlertController (RSSAlertController)
+
+// Made in Mexico
++ (UIAlertController *)addWebSite:(RSSViewController *)rssVC {
+    
+    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Add RSS Channel" message:@"" preferredStyle:UIAlertControllerStyleAlert];
```
Comment (OrientCue) [inline RSReader/Categories/RSSAlertController/UIAlertController+RSSAlertController.m:20]: Please use localized alert messages.

File: RSReader/Categories/RSSAlertController/UIAlertController+RSSAlertController.m [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    __block UITextField *textFieldURL;
+    
+    [alert addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
+        textFieldURL = textField;
+        textFieldURL.placeholder = @"Please enter your URL";
```
Comment (OrientCue) [inline RSReader/Categories/RSSAlertController/UIAlertController+RSSAlertController.m:25]: This text should be localized.

File: RSReader/Categories/RSSAlertController/UIAlertController+RSSAlertController.m [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+#import "RSSViewController.h"
+#import "NSString+URLStringValidator.h"
+#import "UIAlertController+ErrorAlertController.h"
+#import "FeedService.h"
+#import "HtmlParser.h"
+
+@implementation UIAlertController (RSSAlertController)
+
+// Made in Mexico
++ (UIAlertController *)addWebSite:(RSSViewController *)rssVC {
```
Comment (OrientCue) [inline RSReader/Categories/RSSAlertController/UIAlertController+RSSAlertController.m:18]: Missleading method name.
**SRP violation.**

File: RSReader/RSSReader/HtmlParser/HtmlParser.m [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    NSMutableArray *result = [NSMutableArray array];
+    NSString *stringChannel = [[NSString alloc] initWithData:data encoding:NSASCIIStringEncoding];
+    
+
+    NSError *error = nil;
+    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:@"https://[a-z0-9]+.[a-z0-9./]+(rss|feed|xml)"
+                                                                           options:0
+                                                                             error:&error];
+
+    NSTextCheckingResult *match = [regex firstMatchInString:stringChannel
```
Comment (OrientCue) [inline RSReader/RSSReader/HtmlParser/HtmlParser.m:23]: Handling only one feed from website.

File: RSReader/RSSReader/HtmlParser/HtmlParser.m [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+
+@implementation HtmlParser
+
++ (NSArray *)parseHtmlFromData:(NSData *)data {
+    NSMutableArray *result = [NSMutableArray array];
+    NSString *stringChannel = [[NSString alloc] initWithData:data encoding:NSASCIIStringEncoding];
+    
+
+    NSError *error = nil;
+    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:@"https://[a-z0-9]+.[a-z0-9./]+(rss|feed|xml)"
```
Comment (OrientCue) [inline RSReader/RSSReader/HtmlParser/HtmlParser.m:19]: It is possible that feed doesn't match this pattern. Look carefull task description and search here for tag `<link rel="alternate" type="application/rss+xml"`

**Review by OrientCue — CHANGES_REQUESTED** [review]
There are no enhancements from v1.2 (annotation button on cell).
Missed handling website url with http scheme and direct link to feed. 
Please fix comments.

File: RSReader/RSSReader/View Controllers/RSSViewController/RSSViewController.m [hunk 11]
```diff
@@ -0,0 +1,192 @@
...
+        [self.inputChannelUrls removeObjectAtIndex:indexPath.row];
+        [self removeDataFromDocFile:indexPath];
+        dispatch_async(dispatch_get_main_queue(), ^{
+            [self.rssTableView reloadData];
+        });
+    }
+}
+
+- (void)setEditing:(BOOL)editing animated:(BOOL)animated {
+    if (editing && !self.rssTableView.isEditing) {
```
Comment (AndreiChernyshou) [inline RSReader/RSSReader/View Controllers/RSSViewController/RSSViewController.m:121]: It should be done the other way:

Bool isEditing = editing && !self.rssTableView.isEditing;
[self.rssTableView setEditing: isEditing animated: isEditing]

**Review by AndreiChernyshou — APPROVED** [review]

File: RSReader/RSSReader/View Controllers/RSSViewController/RSSViewController.m (hunk 11 above)
Comment (edwardivash) [reply RSReader/RSSReader/View Controllers/RSSViewController/RSSViewController.m]: Fixed

**Review by AndreiChernyshou — APPROVED** [review]

**Pull request merged**